pkg/clientaccess/token_test.go

```diff
@@ -0,0 +1,401 @@
+func newTLSServer(t *testing.T, username, password string, sendWrongCA bool) *httptest.Server {
+	var server *httptest.Server
+	server = httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/v1-k3s/server-bootstrap" {
+			if authUsername, authPassword, ok := r.BasicAuth(); ok != true || authPassword != password || authUsername != username {
+				http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+				return
+			}
+			bootstrapData := &config.ControlRuntimeBootstrap{}
+			w.Header().Set("Content-Type", "application/json")
+			if err := bootstrap.Write(w, bootstrapData); err != nil {
+				t.Errorf("failed to write bootstrap: %v", err)
+			}
+			return
+		}
+
+		if r.URL.Path == "/cacerts" {
+			w.Header().Set("Content-Type", "text/plain")
+			if _, err := w.Write(getServerCA(server)); err != nil {
+				t.Errorf("Failed to write cacerts: %v", err)
+			}
+			return
+		}
+
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+	}))
+
+	// Create new CA cert and key
+	caCert, caKey, err := factory.GenCA()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Generate new server cert; reuse the key from the CA
+	cfg := cert.Config{
+		CommonName:   "localhost",
+		Organization: []string{"testing"},
+		Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		AltNames: cert.AltNames{
+			DNSNames: []string{"localhost"},
+			IPs:      []net.IP{net.IPv4(127, 0, 0, 1)},
+		},
+	}
+	serverCert, err := cert.NewSignedCert(cfg, caKey, caCert, caKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Bind server and CA certs into chain for TLS listener configuration
+	server.TLS = &tls.Config{}
+	server.TLS.Certificates = []tls.Certificate{
+		{Certificate: [][]byte{serverCert.Raw}, Leaf: serverCert, PrivateKey: caKey},
+		{Certificate: [][]byte{caCert.Raw}, Leaf: caCert},
+	}
+
+	if sendWrongCA {
+		// Create new CA cert and key and use that as the CA cert instead of the one that actually signed the server cert
+		badCert, _, err := factory.GenCA()
+		if err != nil {
+			t.Fatal(err)
+		}
+		server.TLS.Certificates[1].Certificate[0] = badCert.Raw
+		server.TLS.Certificates[1].Leaf = badCert
+	}
+
+	server.StartTLS()
+	return server
+}
+
+// getServerCA returns a byte slice containing the PEM encoding of the server's CA certificate
+func getServerCA(server *httptest.Server) []byte {
+	certLen := len(server.TLS.Certificates)
+	return pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: server.TLS.Certificates[certLen-1].Certificate[0]})
+}
+
+// writeServerCA writes the PEM-encoded server certificate to a given path
+func writeServerCA(server *httptest.Server, path string) error {
+	certOut, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer certOut.Close()
+
+	if _, err := certOut.Write(getServerCA(server)); err != nil {
```

Comment:
```suggestion
	return certOut.Write(getServerCA(server))
```

Reply:
I don't think I can ignore one of the return elements, can I?

Reply:
crap. I'm tired. I missed the first value. Ignore!

Reply:
Was hoping there was a shorthand for it that I wasn't aware of lol